### PR TITLE
CFE-3925: Enabled settings environment vars in body agent control via augments

### PR DESCRIPTION
**Pull request by nickanderson**
Supersedes https://github.com/cfengine/masterfiles/pull/2248

### CODE REVIEW

**Comment by nickanderson** [comment]
@flynn1973  can you test this one out? I pulled in your commit from your original PR and then tweaked some things.

**Comment by flynn1973** [comment]
this works
````
/var/cfengine/bin/cf-agent -KI --show-evaluated-vars=environment
...
    Variable name                            Variable value                                               Meta tags                                Comment       
default:control_agent.environment         {"XPG_SUS_ENV=ON"}                                          source=promise                                         
default:def.control_agent_environment_vars  {"XPG_SUS_ENV=ON"}                                          source=augments_file                                 
default:stdlib_common.input[guest_environments] /var/cfengine/inputs/lib/guest_environments.cf               source=promise                                  

````
this throws an error
````
root@aixtest01: /var/cfengine/inputs # /var/cfengine/bin/cf-agent -KI -D debian --show-evaluated-vars=environment
error: Failed to set environment variable '@(def.environment_vars)'. (putenv: Invalid argument)
error: Failed to set environment variable '@(def.environment_vars)'. (putenv: Invalid argument)
...
Variable name                            Variable value                                               Meta tags                                Comment       
default:control_agent.environment         {"@(def.environment_vars)"}                                 source=promise                                         
default:def.control_agent_environment_vars  {"XPG_SUS_ENV=ON"}                                          source=augments_file                                 
default:stdlib_common.input[guest_environments] /var/cfengine/inputs/lib/guest_environments.cf               source=promise

````

**Comment by flynn1973** [comment]
i tinkered around with this snippet...

````
any::
      "control_agent_environment_vars" ->  { "CFE-3925" }
        slist => {
                   ifelse(  isvariable( "def.control_agent_environment_vars" ),
                            $(def.control_agent_environment_vars),
                            " " ),
                   ifelse( "debian", "DEBIAN_FRONTEND=noninteractive", " " ),
                 };
````
but was confronted with totally weird behaviour...

````
root@aixtest01: /var/cfengine/inputs/controls # /var/cfengine/bin/cf-agent -KI -D debian --show-evaluated-vars=environment
...
Variable name                            Variable value                                               Meta tags                                Comment       
default:control_agent.environment         {"@(def.environment_vars)"}                                 source=promise                                         
default:def.control_agent_environment_vars  {"DEBIAN_FRONTEND=noninteractive","DEBIAN_FRONTEND=noninteractive"} source=promise                               
default:stdlib_common.input[guest_environments] /var/cfengine/inputs/lib/guest_environments.cf               source=promise
````
````
root@aixtest01: /var/cfengine/inputs/controls # /var/cfengine/bin/cf-agent -KI  --show-evaluated-vars=environment
...
Variable name                            Variable value                                               Meta tags                                Comment       
default:control_agent.environment         {" "," "}                                                   source=promise                                         
default:def.control_agent_environment_vars  {" "," "}                                                   source=promise                                       
default:stdlib_common.input[guest_environments] /var/cfengine/inputs/lib/guest_environments.cf               source=promise
````

**Comment by nickanderson** [comment]
OK, so agent errors if environment is an empty list or if the list does not contain `KEY=VALUE` pairs. I filed CFE-3927 to allow setting an empty list and added a commit to re-factor and protect against setting environment in body agent control when the variable passes some validation. Validation currently is variable defined, variable not 0 length, each item in list is key=value, and for 3.18.0 and newer hosts we check that the variable type is slist or data array.

Can you give it a go again @flynn1973 ?

**Comment by flynn1973** [comment]
var configured in def.json is not added to control_agent.environment 
````
/var/cfengine/bin/cf-agent -KI  --show-evaluated-vars=environment
...
Variable name                            Variable value                                               Meta tags                                Comment       
default:def.control_agent_environment_vars  {"XPG_SUS_ENV=ON"}                                          source=augments_file                                 
default:stdlib_common.input[guest_environments] /var/cfengine/inputs/lib/guest_environments.cf               source=promise 
````
also not as expected...
````
/var/cfengine/bin/cf-agent -KI -D debian --show-evaluated-vars=environment
...
Variable name                            Variable value                                               Meta tags                                Comment       
default:control_agent.environment         {"DEBIAN_FRONTEND=noninteractive"}                          source=promise                                         
default:def.control_agent_environment_vars  {"XPG_SUS_ENV=ON"}                                          source=augments_file                                 
default:def.environment_vars              {"DEBIAN_FRONTEND=noninteractive"}                          source=promise                           Environment variables of the agent process. The values
                    of environment variables are inherited by child commands.
default:stdlib_common.input[guest_environments] /var/cfengine/inputs/lib/guest_environments.cf               source=promise                
````
expected result
````
Variable name                            Variable value                                               Meta tags                                Comment       
default:control_agent.environment         {"DEBIAN_FRONTEND=noninteractive", "XPG_SUS_ENV=ON"}                          source=promise                                         
default:def.control_agent_environment_vars  {"XPG_SUS_ENV=ON"}                                          source=augments_file                                 
default:def.environment_vars              {"DEBIAN_FRONTEND=noninteractive"}                          source=promise                           Environment variables of the agent process. The values
                    of environment variables are inherited by child commands.
default:stdlib_common.input[guest_environments] /var/cfengine/inputs/lib/guest_environments.cf               source=promise                
````

**Comment by nickanderson** [comment]
Hummm, what version of the agent are you using? This is the behavior that I see:

```
[root@hub masterfiles]# cf-agent --version
CFEngine Core 3.18.1
```

As expected when there is no default (EL host) and not set via `def.json:

```
[root@hub masterfiles]# cat /var/cfengine/inputs/def.json
cat: /var/cfengine/inputs/def.json: No such file or directory
[root@hub masterfiles]# cf-agent -K --show-evaluated-vars="control_agent\.environment|control_agent_environment_vars"
Variable name                            Variable value                                               Meta tags                                Comment   
```

As expected when not defined via `def.json` and explicitly defining the `debian` class, the policy default is picked up.

```
[root@hub masterfiles]# cat /var/cfengine/inputs/def.json
cat: /var/cfengine/inputs/def.json: No such file or directory
[root@hub masterfiles]# cf-agent -K --define debian --show-evaluated-vars="control_agent\.environment|control_agent_environment_vars"
Variable name                            Variable value                                               Meta tags                                Comment                                 
default:control_agent.environment         {"DEBIAN_FRONTEND=noninteractive"}                          source=promise                                                                   
default:def.control_agent_environment_vars  {"DEBIAN_FRONTEND=noninteractive"}                          source=promise                           Set default environment variables for using Debian non-interactively
```

As expected when `def.control_agent_environment_vars` is set via `def.json`:

```
[root@hub masterfiles]# cat /var/cfengine/inputs/def.json 
{
  "vars": {
    "control_agent_environment_vars": [ "BEST_CONFIG_MGMT=CFEngine" ]
  }
}
[root@hub masterfiles]# cf-agent -K --show-evaluated-vars="control_agent\.environment|control_agent_environment_vars"
Variable name                            Variable value                                               Meta tags                                Comment                                 
default:control_agent.environment         {"BEST_CONFIG_MGMT=CFEngine"}                               source=promise                                                                   
default:def.control_agent_environment_vars  {"BEST_CONFIG_MGMT=CFEngine"}                               source=augments_file                  
```

As expected, when `def.control_agent_environment_vars` is set via `def.json` and there is a class defined where there would be a policy default, the policy default is not applied since the value is set in `def.json`:

```
[root@hub masterfiles]# cat /var/cfengine/inputs/def.json
{
  "vars": {
    "control_agent_environment_vars": [ "BEST_CONFIG_MGMT=CFEngine" ]
  }
}
[root@hub masterfiles]# cf-agent -K --define debian --show-evaluated-vars="control_agent\.environment|control_agent_environment_vars"
Variable name                            Variable value                                               Meta tags                                Comment                                 
default:control_agent.environment         {"BEST_CONFIG_MGMT=CFEngine"}                               source=promise                                                                   
default:def.control_agent_environment_vars  {"BEST_CONFIG_MGMT=CFEngine"}                               source=augments_file           
```

**Comment by nickanderson** [comment]
> ok, until now, it was my belief that whole rumble was about the following:
this debian attribute is a masterfiles provided default value which should always be in control_agent.environment. if someone is setting custom control_agent_environment_vars in def.json, this value should be ADDED to the list of vars -> control_agent.environment
the actual situation does not expand but only allows one item in this list, which is not desired.

Yeah, no, this implementation allows you to set the environment vars, not tack on to the defaults.

However, more than 1 Key value should definitely work here.

```
[root@hub masterfiles]# cat /var/cfengine/inputs/def.json
{
  "vars": {
    "control_agent_environment_vars": [ "BEST_CONFIG_MGMT=CFEngine", "DEBIAN_FRONTEND=noninteractive", "XPG_SUS_ENV=ON"  ]
  }
}
[root@hub masterfiles]# cf-agent -K --define debian --show-evaluated-vars="control_agent\.environment|control_agent_environment_vars"
Variable name                            Variable value                                               Meta tags                                Comment                                 
default:control_agent.environment         {"BEST_CONFIG_MGMT=CFEngine","DEBIAN_FRONTEND=noninteractive","XPG_SUS_ENV=ON"} source=promise                                                                   
default:def.control_agent_environment_vars  {"BEST_CONFIG_MGMT=CFEngine","DEBIAN_FRONTEND=noninteractive","XPG_SUS_ENV=ON"} source=augments_file    
```

You can get different settings for different hosts by leveraging the augments key:

```
[root@hub masterfiles]# cat /var/cfengine/inputs/def.json 
{
  "vars": {
    "control_agent_environment_vars": [ "BEST_CONFIG_MGMT=CFEngine", "DEBIAN_FRONTEND=noninteractive", "XPG_SUS_ENV=ON"  ]
  },
  "augments": [ "$(sys.flavor).def.json" ]
}
[root@hub masterfiles]# cat /var/cfengine/inputs/centos_7.def.json 
{
  "vars": {
    "control_agent_environment_vars": [ "CENTOS_7=Something Centos Specific" ]
  }
}
[root@hub masterfiles]# cf-agent -K --define debian --show-evaluated-vars="control_agent\.environment|control_agent_environment_vars"
Variable name                            Variable value                                               Meta tags                                Comment                                 
default:control_agent.environment         {"CENTOS_7=Something Centos Specific"}                      source=promise                                                                   
default:def.control_agent_environment_vars  {"CENTOS_7=Something Centos Specific"}                      source=augments_file         
```

 Or by using the host-specific-data. Extending whatever the default are would be a separate feature I think, a good one to have, but I think that should be a different ticket and pr which would be great if you want to work on after this one.

**Comment by flynn1973** [comment]
well ok , but the main problem was to not overwrite the debian guys apt var with augments values :-), not sure how to proceed here. should i create another ticket?

**Comment by nickanderson** [comment]
> well ok , but the main problem was to not overwrite the debian guys apt var with augments values :-), not sure how to proceed here. should i create another ticket

Yeah, I think a different ticket would make sense. 

This implementation lets you override the defautls. You might want to extend the defaults and you can get that behavior by stating the defaults along with your others.

The implementation in #2248 would not take effect at all on debian hosts since there was a debian specific setting right in body agent control. https://github.com/flynn1973/masterfiles/blob/f4e7031667493269c506374a4b97fddb13349dcb/controls/cf_agent.cf#L60-L62. So, if you had set `mpf_control_agent_environment_vars` for example on debian hosts the changes would not have taken effect.

**Comment by flynn1973** [comment]
i am totally ok with this implemenation as it allows me to set  vars  for my aix needs..will create another ticket to keep the debian crowd happy. another option would be  to make it clear in the docs that needed vars must be set in augments from now on.

**Comment by nickanderson** [comment]
> i am totally ok with this implemenation as it allows me to set vars for my aix needs..will create another ticket to keep the debian crowd happy. another option would be to make it clear in the docs that needed vars must be set in augments from now on, or set it in the apt package module.

Yeah, this pr has this:


```
### Environment Variables

Environment variables that should be inherited by child commands can be set using `def.control_agent_environment_vars`. The MPF tries to provide reasonable defaults but this is configurable via Augments, for example:

```json
{
  "vars": {
    "control_agent_environment_vars":
      [ "DEBIAN_FRONTEND=noninteractive",
        "XPG_SUS_ENV=ON" ]
  }
}
```

**Notes:**

* A simple augments as shown above applies to *all* hosts. Consider using the
  augments key or host specific data if you want to set environment variables
  differently across different sets of hosts. The value set via Augments takes
  precedence over policy defaults, so be sure to take that into account when
  configuring.
```

Do you think it needs more?

I don't think that we will ever be able to completely replace the power of just editing the policy for your needs but we can sure keep trying :P

**Comment by nickanderson** [comment]
@flynn1973 it really isn't much to tweak to add the append/extend the defaults behavior ...

I am on an el7 host with "extra" environment vars defined in def.json. I define debian and get my extra plus the debian default.

```
[root@hub masterfiles]# cat /var/cfengine/inputs/def.json 
{
  "vars": {
    "control_agent_environment_vars_extra": [ "BEST_CONFIG_MGMT=CFEngine" ]
  }
}
[root@hub masterfiles]# cf-agent -K --define debian --show-evaluated-vars="control_agent\.environment|control_agent_environment_vars"
Variable name                            Variable value                                               Meta tags                                Comment                                 
default:control_agent.environment         {"DEBIAN_FRONTEND=noninteractive","BEST_CONFIG_MGMT=CFEngine"} source=promise                                                                   
default:def.control_agent_environment_vars  {"DEBIAN_FRONTEND=noninteractive","BEST_CONFIG_MGMT=CFEngine"} source=promise                                                                   
default:def.control_agent_environment_vars_default  {"DEBIAN_FRONTEND=noninteractive"}                          source=promise                           Set default environment variables for using Debian non-interactively
default:def.control_agent_environment_vars_extra  {"BEST_CONFIG_MGMT=CFEngine"}                               source=augments_file         
```

I do a normal run, so no debain, i just get my extra vars used since there is no default for el.

```                                                    
[root@hub masterfiles]# cf-agent -K --show-evaluated-vars="control_agent\.environment|control_agent_environment_vars"
Variable name                            Variable value                                               Meta tags                                Comment                                 
default:control_agent.environment         {"BEST_CONFIG_MGMT=CFEngine"}                               source=promise                                                                   
default:def.control_agent_environment_vars  {"BEST_CONFIG_MGMT=CFEngine"}                               source=promise                                                                   
default:def.control_agent_environment_vars_extra  {"BEST_CONFIG_MGMT=CFEngine"}                               source=augments_file                   
```

I get rid of my extra vars and run with no extra classes getting nothing for environemnt vars:

```                                          
[root@hub masterfiles]# mv /var/cfengine/inputs/def.json{,.bak}
[root@hub masterfiles]# cf-agent -K --show-evaluated-vars="control_agent\.environment|control_agent_environment_vars"
Variable name                            Variable value                                               Meta tags                                Comment                                 
```

I run with debian defined and I get just the debian defaults.

```
[root@hub masterfiles]# cf-agent -K --define debian --show-evaluated-vars="control_agent\.environment|control_agent_environment_vars"
Variable name                            Variable value                                               Meta tags                                Comment                                 
default:control_agent.environment         {"DEBIAN_FRONTEND=noninteractive"}                          source=promise                                                                   
default:def.control_agent_environment_vars  {"DEBIAN_FRONTEND=noninteractive"}                          source=promise                                                                   
default:def.control_agent_environment_vars_default  {"DEBIAN_FRONTEND=noninteractive"}                          source=promise                           Set default environment variables for using Debian non-interactively
[root@hub masterfiles]# 
```

I figured it would result in more understandable variable names to change this now before it went in the first way.

Please review the behavior and let me know if you approve.

**Comment by flynn1973** [comment]
bingo! works perfectly....

**Comment by flynn1973** [comment]
just to be sure i tested on a "real" debian system...

````
root@t560:/var/cfengine/inputs/controls# cf-agent --version
CFEngine Core 3.18.1
root@t560:/var/cfengine/inputs/controls# cf-agent -K --show-evaluated-vars="control_agent\.environment|control_agent_environment_vars"
/var/cfengine/inputs/controls/def.cf:272:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:276:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:280:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:284:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:291:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:296:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:301:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:309:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:313:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:321:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:331:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:337:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:366:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:402:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:408:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:414:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:420:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:446:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:453:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:458:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:464:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:486:0: error: Common attribute 'comment' not implemented for custom promises (any)
/var/cfengine/inputs/controls/def.cf:491:0: error: Common attribute 'comment' not implemented for custom promises (any)
   error: Unresolved variable '@(def.augments_inputs)' in input list, cannot parse
   error: There are syntax errors in policy files
   error: Policy failed validation with command '"/var/cfengine/bin/cf-promises" -c "/var/cfengine/inputs/promises.cf"'
   error: Failsafe condition triggered. Interactive session detected, skipping failsafe.cf execution.
   error: Error reading CFEngine policy. Exiting...

````

**Comment by flynn1973** [comment]
when commenting out the "any::" at https://github.com/nickanderson/masterfiles/blob/CFE-3925/master/controls/def.cf#L271 it works. i dont have the slightest clue why...this is way beyond my skillset.
````
root@t560:/var/cfengine/inputs/controls# cf-agent -KI --show-evaluated-vars="control_agent\.environment|control_agent_environment_vars"
Variable name                            Variable value                                               Meta tags                                Comment                                 
default:control_agent.environment         {"DEBIAN_FRONTEND=noninteractive","BEST_CONFIG_MGMT=CFEngine"} source=promise                                                                   
default:def.control_agent_environment_vars  {"DEBIAN_FRONTEND=noninteractive","BEST_CONFIG_MGMT=CFEngine"} source=promise                                                                   
default:def.control_agent_environment_vars_default  {"DEBIAN_FRONTEND=noninteractive"}                          source=promise                           Set default environment variables for using Debian non-interactively
default:def.control_agent_environment_vars_extra  {"BEST_CONFIG_MGMT=CFEngine"}                               source=augments_file 
````

**Comment by nickanderson** [comment]
> when commenting out the "any::" at https://github.com/nickanderson/masterfiles/blob/CFE-3925/master/controls/def.cf#L271 it works. i dont have the slightest clue why...this is way beyond my skillset.
> 
> ```
> root@t560:/var/cfengine/inputs/controls# cf-agent -KI --show-evaluated-vars="control_agent\.environment|control_agent_environment_vars"
> Variable name                            Variable value                                               Meta tags                                Comment                                 
> default:control_agent.environment         {"DEBIAN_FRONTEND=noninteractive","BEST_CONFIG_MGMT=CFEngine"} source=promise                                                                   
> default:def.control_agent_environment_vars  {"DEBIAN_FRONTEND=noninteractive","BEST_CONFIG_MGMT=CFEngine"} source=promise                                                                   
> default:def.control_agent_environment_vars_default  {"DEBIAN_FRONTEND=noninteractive"}                          source=promise                           Set default environment variables for using Debian non-interactively
> default:def.control_agent_environment_vars_extra  {"BEST_CONFIG_MGMT=CFEngine"}                               source=augments_file 
> ```

That error is because of a typo I made:

https://github.com/cfengine/masterfiles/pull/2249/files#diff-6ba06b25655893d02a8da5a8dea0071f1272881c7d418749bd9079f87ddd11faR271

only a single `:` after any. Can you double check?

**Comment by flynn1973** [comment]
god :-) you are right...everythin ok then.

**Comment by nickanderson** [comment]
Squashed.

**Comment by nickanderson** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7603/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7603/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7603/

File: MPF.md [hunk 1]
```diff
@@ -1392,6 +1392,44 @@ the watchdog will not be active.
...
+  "vars": {
+    "control_agent_environment_vars_extra": [ "XPG_SUS_ENV=ON" ]
+  }
+}
+```
+
+**Notes:**
+
+* Simple augments as shown above apply to *all* hosts. Consider using the
+  augments key or host specific data if you want to set environment variables
```
Comment (craigcomstock) [inline MPF.md:1424]: can you link to what "augments key" is? and also link to "host specific data"?

File: controls/def.cf [hunk 2]
```diff
@@ -226,18 +226,60 @@ bundle common def
         int => "200",
         if => not( isvariable( "control_server_maxconnections" ) );
 
+      #+begin_src def.control_agent_environment_vars
+      # This configures environment_vars in body agent control
+
+      # It's configurable without having to modify policy, so the default values
+      # are only applied if the variable is not already defined (via augments)
```
Comment (craigcomstock) [inline controls/def.cf:233]: period at end of sentence

Comment (craigcomstock) [reply controls/def.cf:233]: Remove org mode "#+begin_src"?

File: controls/def.cf [hunk 3]
```diff
@@ -226,18 +226,60 @@ bundle common def
...
+      # Resolve the final state for environment vars if there is a default var
+      # or if there is a user extra and merge the defined entities
+      "control_agent_environment_vars" ->  { "CFE-3925" }
+        slist => { @(def.control_agent_environment_vars_default),
+                   @(def.control_agent_environment_vars_extra) },
+        policy => "ifdefined",
+        if => or( isvariable( "def.control_agent_environment_vars_default" ), # Protect against defining
+                  isvariable( "def.control_agent_environment_vars_extra" ));  # an empty variable
+
+      #+end_src Agent Environment Variables
```
Comment (craigcomstock) [inline controls/def.cf:282]: This is a cf file, why is org stuff in here? ;)

**Review by craigcomstock — COMMENTED** [review]
a few comments.

File: controls/def.cf (hunk 3 above)
Comment (nickanderson) [reply controls/def.cf:282]: https://github.com/cfengine/masterfiles/pull/2249#discussion_r812218815

File: MPF.md (hunk 1 above)
Comment (nickanderson) [reply MPF.md:1424]: +1, I will use the name targets, not full urls since this page is rendered in docs.

**Comment by nickanderson** [comment]
> a few comments.

Added links, period. Leaving org formatted comments (ref: https://github.com/cfengine/masterfiles/pull/2249#discussion_r812218815). Ready for re-review.

**Review by vpodzime — APPROVED** [review]
ACK

**Pull request merged**